File: iOS/TCEffectPlayerKit/Classes/TUIExtension/TUIEffectPlayerService.m
//
//  TUIEffectPlayerService.m
//  TCEffectPlayerKit
//
//  Created by krabyu on 2024/7/18.
//

#import "TUIEffectPlayerService.h"
#import "TUICore/TUICore.h"
#import "TCConstant.h"
#import <TCEffectPlayer/TCEffectAnimView.h>
#import <TCMediaX/TCMediaX.h>

@interface TUIEffectPlayerService() <TUIServiceProtocol,TCEPAnimViewDelegate>
@property (nonatomic, copy) TUICallServiceResultCallback callback;
@end

@implementation TUIEffectPlayerService
+ (void)load {
    [TUICore registerService:KEY_SERVICE_NAME object: [TUIEffectPlayerService sharedInstance]];
}

+ (instancetype) sharedInstance {
    static TUIEffectPlayerService* service;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        service = [[TUIEffectPlayerService alloc] init];
    });
    return service;
}

- (id)onCall:(NSString *)method param:(nullable NSDictionary *)param resultCallback:(TUICallServiceResultCallback)resultCallback {
    if ([method isEqualToString: KEY_METHOD_SET_LICENSE]) {
        NSString *licenseUrl = param[KEY_PARAM_LICENSE_URL];
        NSString *licenseKey = param[KEY_PARAM_LICENSE_KEY];
        if (licenseUrl && licenseKey ) {
            [[TCMediaXBase getInstance] setLicenceURL:licenseUrl key:licenseKey];
        }
    } else if ([method isEqualToString:KEY_METHOD_START_PLAY]) {
        NSString * url = param[KEY_PARAM_PLAY_URL];
        UIView* view = param[KEY_PARAM_VIEW];
        if ([view isKindOfClass: [TCEffectAnimView class]]) {
            TCEffectAnimView * animationView = (TCEffectAnimView* )view;
            [animationView setEffectPlayerDelegate:self];
            [animationView startPlay:url];
            self.callback = resultCallback;
        }
    } else if ([method isEqualToString:KEY_METHOD_STOP_PLAY]) {
        UIView* view = param[KEY_PARAM_VIEW];
        if ([view isKindOfClass: [TCEffectAnimView class]]) {
            TCEffectAnimView * animationView = (TCEffectAnimView* )view;
            [animationView setEffectPlayerDelegate:NULL];
            [animationView stopPlay];
        }
    }
    return nil;
}
    
#pragma mark - TCEPAnimViewDelegate
-(void) onPlayEvent:(ITCEffectPlayer *)player event:(int)EvtID withParam:(NSDictionary *)param {
    int event_play_stop = 2006;
    if (EvtID == event_play_stop && self.callback) {
        self.callback(0, @"play finished", @{@"":@""});
    }
}

@end

File: Android/tceffectplayerkit/src/main/java/com/tencent/tcmediax/tceffectplayerkit/core/TUIEffectPlayerService.java
package com.tencent.tcmediax.tceffectplayerkit.core;

import android.content.Context;
import android.os.Bundle;
import android.text.TextUtils;

import com.tencent.qcloud.tuicore.interfaces.ITUIService;
import com.tencent.qcloud.tuicore.interfaces.TUIServiceCallback;
import com.tencent.tcmediax.api.TCMediaXBase;
import com.tencent.tcmediax.tceffectplayerkit.Constants;
import com.tencent.tcmediax.tceffectplayer.api.TCEffectAnimView;

import java.util.Map;

public class TUIEffectPlayerService implements ITUIService {

    @Override
    public Object onCall(String method, Map<String, Object> param, TUIServiceCallback callback) {
        if (param != null && TextUtils.equals(Constants.KEY_METHOD_START_PLAY, method)) {
            String url = (String) param.get(Constants.KEY_PARAM_PLAY_URL);
            Object view = param.get(Constants.KEY_PARAM_VIEW);
            if (view instanceof TCEffectAnimView) {
                TCEffectAnimView animView = (TCEffectAnimView) view;
                animView.setPlayListener(new AnimPlayListener(callback));
                animView.startPlay(url);
            }
        } else if (param != null && TextUtils.equals(Constants.KEY_METHOD_STOP_PLAY, method)) {
            Object view = param.get(Constants.KEY_PARAM_VIEW);
            Boolean clearLastFrame = (Boolean) param.get(Constants.KEY_PARAM_CLEAR_LAST_FRAME);
            if (view instanceof TCEffectAnimView) {
                TCEffectAnimView animView = (TCEffectAnimView) view;
                animView.setPlayListener(null);
                animView.stopPlay(Boolean.TRUE.equals(clearLastFrame));
            }
        } else if (param != null && TextUtils.equals(Constants.KEY_METHOD_SET_LICENSE, method)) {
            String url = (String) param.get(Constants.KEY_PARAM_LICENSE_URL);
            String key = (String) param.get(Constants.KEY_PARAM_LICENSE_KEY);
            Context context = (Context) param.get(Constants.KEY_PARAM_CONTEXT);
            TCMediaXBase.getInstance().setLicense(context, url, key, (i, s) -> {
                if (callback != null) {
                    callback.onServiceCallback(i, s, null);
                }
            });
        }
        return null;
    }

    private static final class AnimPlayListener implements TCEffectAnimView.IAnimPlayListener {

        private final TUIServiceCallback mCallback;

        public AnimPlayListener(TUIServiceCallback callback) {
            mCallback = callback;
        }

        @Override
        public void onPlayStart() {

        }

        @Override
        public void onPlayEnd() {
            if (mCallback != null) {
                mCallback.onServiceCallback(0, "", null);
            }
        }

        @Override
        public void onPlayError(int i) {
            if (mCallback != null) {
                mCallback.onServiceCallback(i, "", null);
            }
        }

        @Override
        public void onPlayEvent(int i, Bundle bundle) {

        }
    }
}
